rohmu/object_storage/s3.py

```diff
@@ -486,6 +488,26 @@ def get_file_size(self, key: str) -> int:
             else:
                 raise StorageError(f"File size lookup failed for {path}") from ex
 
+    def calculate_chunks_and_chunk_size(self, size: Optional[int]) -> tuple[int, int]:
+        """Calculate the number of chunks and chunk size for multipart upload.
+
+        If sizes provided self.default_multipart_chunk_size wil be used as first attempt,
+        if number of chunks is greater than S3_MAX_NUM_PARTS_PER_UPLOAD, chunk size will be doubled,
+        until the number of chunks is less than S3_MAX_NUM_PARTS_PER_UPLOAD.
+        """
+        if size is None:
+            return 1, self.default_multipart_chunk_size
+        chunks = math.ceil(size / self.default_multipart_chunk_size)
+        chunk_size = self.default_multipart_chunk_size
+
+        while chunks > S3_MAX_NUM_PARTS_PER_UPLOAD:
+            chunk_size *= 2
+            if chunk_size > S3_MAX_PART_SIZE_BYTES:
+                raise StorageError(f"Cannot upload a file of size {size}. Chunk size {chunk_size} is too big for each part of multipart upload.")
+            chunks = math.ceil(size / chunk_size)
```

Comment:
Can't we just do:

```suggestion
        if chunks > S3_MAX_NUM_PARTS_PER_UPLOAD:
            chunk_size = math.ceil(size / S3_MAX_NUM_PARTS_PER_UPLOAD)
            if chunk_size > S3_MAX_PART_SIZE_BYTES:
                raise StorageError(f"Cannot upload a file of size {size}. Chunk size {chunk_size} is too big for each part of multipart upload.")
            chunks = math.ceil(size / chunk_size)
```

?

This would find the smallest chunk size that fits exactly in the maximum number of chunks without having to loop. The cons is that the chunk size won't be a "round number".

Reply:
good idea, applied it. I don't think there's an obvious advantage between doing slightly bigger chunk size vs more chunks with smaller size, so for simplicity I opted for your solution. we may re-evaluate this later on.